### PR TITLE
[Results] Make sorting and filtering not eagerly create intermediary views

### LINKED ISSUE #1160: Skip sorting when calling count on an RLMResults that hasn't been constructed yet
Continuing the discussion from [this thread] (https://groups.google.com/forum/#!topic/realm-cocoa/3zWhXrCcIic) on the mailing list, I was made aware of how the count property on RLMResults works.

From doing a little more benchmarking, it seems to me that sorting my large query by a single NSDate property is the most expensive operation being done by Realm. Further, it also looks like, Realm is doing the sorting even for count-only queries. I came to this conclusion by testing the following 2 queries.

    [[[Article allObjectsInRealm:realm] sortedResultsUsingProperty:@"date" ascending:YES] count];
    [[Article allObjectsInRealm:realm] count];

The difference was significant. So, if this is indeed the case, do you guys think, not doing sorting on count-only queries is something Realm should handle internally?

### PR DESCRIPTION
**Pull request by segiddins**
Closes #1160 

\c @tgoyne

### CODE REVIEW

**Comment by jpsim** [comment]
Please add a changelog entry and a test confirming that the tableview isn't materialized when checking the count of a sorted RLMResults.

File: Realm/RLMResults.mm [hunk 1]
```diff
@@ -64,7 +65,7 @@ + (instancetype)resultsWithObjectClassName:(NSString *)objectClassName
     ar->_objectClassName = objectClassName;
     ar->_viewCreated = NO;
     ar->_backingQuery = move(query);
-    ar->_sortOrder = sorter;
+    ar->_sortOrder = std::move(sorter);
```
Comment (tgoyne) [inline Realm/RLMResults.mm]: `move` doesn't do anything here (can't move from a const reference)

File: Realm/RLMResults.mm [hunk 2]
```diff
@@ -119,6 +120,14 @@ static inline void RLMResultsValidateInWriteTransaction(__unsafe_unretained RLMR
     }
 }
 
+static std::unique_ptr<RowIndexes::Sorter> RLMSorterFromDescriptors(RLMObjectSchema *schema, NSArray *descriptors)
```
Comment (tgoyne) [inline Realm/RLMResults.mm]: This shouldn't be heap-allocating the sorter.

File: Realm/RLMResults.mm [hunk 3]
```diff
@@ -278,14 +287,14 @@ - (RLMResults *)objectsWhere:(NSString *)predicateFormat args:(va_list)args {
...
-    RLMResultsValidate(self);
+    RLMCheckThread(_realm);
 
     // copy array and apply new predicate creating a new query and view
     auto query = [self cloneQuery];
     RLMUpdateQueryWithPredicate(query.get(), predicate, _realm.schema, _realm.schema[self.objectClassName]);
     return [RLMResults resultsWithObjectClassName:self.objectClassName
                                             query:move(query)
-                                             sort:_backingView.m_sorting_predicate
+                                             sort:std::move(_sortOrder)
```
Comment (tgoyne) [inline Realm/RLMResults.mm]: Don't use `move` here. The Sorter needs to be copied into the new `RLMResults`, not moved from the current one into the new one.

**Comment by tgoyne** [comment]
:+1: 

CI failure appears to just be a CI issue.

**Pull request merged**